### PR TITLE
PMUI 8 user registration

### PR DESCRIPTION
**Pull request by keya5406**
# Branch Description:
This branch focuses on implementing the user registration process.

Key tasks include:
- Creating a signup form with fields for email and master password.
- Validating email format and enforcing strong password requirements.
- Generating a salt and hashing the master password using PBKDF2 for secure storage.
- Sending the hashed password and salt to the backend.
- Handling backend responses for successful or failed registration.
- Thorough testing and validation of the registration flow to ensure proper functionality.

### CODE REVIEW

File: src/components/SignupForm.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            <h2 className="text-2xl font-bold mb-6 text-center">Sign Up</h2>
+
+            <div className="mb-4">
+                <label
+                    htmlFor="email"
+                    className="block text-sm font-medium text-gray-700 mb-1"
+                >
+                    Email:
+                </label>
+                <input
```
Comment (chintal31) [inline src/components/SignupForm.js]: **Refactor Form Components for Reusability**

I would suggest to create a seperate components for input fields and button. For example, 
1. Email.js – A component dedicated to the email input field.
2. Password.js – A component for the password input field.
3. Button.js – A reusable button component that takes a text prop (e.g., "Sign Up", "Login").

```
<form ....>
        <Email>
        <Password>
        <Button text="Sign Up">
  </form>

```

Reasons for the Refactor:

- Reusability: By separating the components, we can easily reuse them across different forms (e.g., login, sign-up, reset password).
- Consistency: Centralized management of styles and behavior (e.g., if we need to change the button's font or color, updating it in one place will reflect across all forms).
- Maintainability: This structure makes the code more modular and easier to manage as the application grows.

**Review by chintal31 — CHANGES_REQUESTED** [review]

**Comment by chintal31** [comment]
Also,
1) can you merge in the main branch so that this branch does not depict the changes of the previous branch?
2) can you update the description of the PR. For example, what is the PR for, what changes does it bring, etc. You can take help of ChatGpt or similar AI tools to help you provide the template for description.

Thanks!

File: src/components/SignupForm.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        if (Object.keys(newErrors).length > 0) {
+            setError(newErrors);
+            return;
+        }
+        const salt = generateSalt();
+        const hashedPassword = hashPassword(masterPassword, salt);
+
+        console.log("sending");
+        try {
+            const response = await axios.post('http://localhost:8080/api/register', {
```
Comment (chintal31) [inline src/components/SignupForm.js]: Here, the url `http://localhost:8080` should come from an env variable (you can check best practices to build a URL online).

File: src/components/SignupForm.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                email: email,
+                hashedPassword: hashedPassword,
+                salt: salt,
+
+            }, {
+                headers: {
+                    'Content-Type': 'application/json',
+                    'Authorization': 'Bearer YOUR_ACCESS_TOKEN'
```
Comment (chintal31) [inline src/components/SignupForm.js]: this header `Authorization: ...` is not required for now. We may implement this later but for now you could remove it.

File: src/components/SignupForm.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                }
+            });
+            // console.log('Email:', email);
+            // console.log('Generated Salt:', salt);
+            // console.log("Password:", masterPassword);
+            // console.log('Hashed Password:', hashedPassword);
+            console.log('Response:', response.data);
+
+            navigate('/CredentialForm');
```
Comment (chintal31) [inline src/components/SignupForm.js]: On successfull signup, the user should navigate to /login page.

**Review by chintal31 — COMMENTED** [review]
Here's how I managed cors error:

- Run the backend server on port 3000 
    - Check backend readme on how to run it incase you need reference, I ran first `gradle bootImage` and then `docker compose up`
    - Make sure you pull the branch which contains api/register feature and start the backend.
- Add a proxy field to package.json
-           { "proxy" : "http://localhost:3000" }
- start the react app, fill signup form, submit and should work   

If you have any questions, feel free to ping me.

**Review by chintal31 — CHANGES_REQUESTED** [review]
Well done implementing the signup functionality! I like the form with shadow, the button, its colors & font. 🆒 👍 

Here are a few improvements:
a) Add loader symbol after clicking any button till response is received from backend.
b) I signed up but did not navigate to login page. Can you check the reason?

After these two improvements, I’d like to suggest a more industry-standard way of structuring the code to improve maintainability, readability, and scalability. Below are my observations and recommendations:  

---

### **1. Single Responsibility Principle**  
Currently, the `SignupForm` component handles multiple responsibilities:  
- UI rendering  
- State management  
- Validation logic  
- API calls  
- Cryptographic operations  

This makes the code harder to test, debug, and scale. To adhere to the single-responsibility principle, we can break it down into smaller, focused modules.  

---

### **2. Proposed Code Structure**  
Here’s how the code can be modularized:  

1. **Component (`SignupForm.tsx`)**  
   The `SignupForm` component should focus only on rendering the form and sub-components (e.g., `Email`, `Password`, `Button`) and delegate the state and logic to a custom hook (`useSignupForm`).  

2. **Custom Hook (`useSignupForm.ts`)**  
   The hook manages form state, handles input changes, and executes the signup flow. It keeps the form logic isolated from the UI, making it easier to test independently.  

3. **API Integration (`api.ts`)**  
   Extract API calls into a separate file to isolate external dependencies. This helps in mocking and testing API interactions separately.  

4. **Utility Functions (`cryptoUtils.ts`)**  
   Move cryptographic operations (`generateSalt`, `hashPassword`) and validation logic (`isValidEmail`, `passwordPattern`) to a utility file. These can be reused across other parts of the application.  

---

### **3. Advantages of This Approach**  
- **Scalability**: Adding new features like a password strength indicator or additional validation becomes straightforward.  
- **Reusability**: Common functionality like email validation or hashing can be reused in other forms (e.g., login or password reset).  
- **Testability**: Each module (UI, logic, utilities, and API) can be tested independently.  
- **Readability**: By separating concerns, it becomes easier for developers to understand and maintain the code.  

---

### **4. Example Implementation**  
You can refer to the following structure as an example:  
- `SignupForm.tsx`: Focuses on rendering the form and integrating sub-components.  
- `useSignupForm.ts`: Centralizes state management and business logic.  
- `api.ts`: Handles API-related code.  
- `cryptoUtils.ts`: Contains utility functions for cryptographic and validation operations.

**Comment by keya5406** [comment]
> Well done implementing the signup functionality! I like the form with shadow, the button, its colors & font. 🆒 👍
> 
> Here are a few improvements: a) Add loader symbol after clicking any button till response is received from backend. b) I signed up but did not navigate to login page. Can you check the reason?
> 
> After these two improvements, I’d like to suggest a more industry-standard way of structuring the code to improve maintainability, readability, and scalability. Below are my observations and recommendations:
> 
> ### **1. Single Responsibility Principle**
> Currently, the `SignupForm` component handles multiple responsibilities:
> 
> * UI rendering
> * State management
> * Validation logic
> * API calls
> * Cryptographic operations
> 
> This makes the code harder to test, debug, and scale. To adhere to the single-responsibility principle, we can break it down into smaller, focused modules.
> 
> ### **2. Proposed Code Structure**
> Here’s how the code can be modularized:
> 
> 1. **Component (`SignupForm.tsx`)**
>    The `SignupForm` component should focus only on rendering the form and sub-components (e.g., `Email`, `Password`, `Button`) and delegate the state and logic to a custom hook (`useSignupForm`).
> 2. **Custom Hook (`useSignupForm.ts`)**
>    The hook manages form state, handles input changes, and executes the signup flow. It keeps the form logic isolated from the UI, making it easier to test independently.
> 3. **API Integration (`api.ts`)**
>    Extract API calls into a separate file to isolate external dependencies. This helps in mocking and testing API interactions separately.
> 4. **Utility Functions (`cryptoUtils.ts`)**
>    Move cryptographic operations (`generateSalt`, `hashPassword`) and validation logic (`isValidEmail`, `passwordPattern`) to a utility file. These can be reused across other parts of the application.
> 
> ### **3. Advantages of This Approach**
> * **Scalability**: Adding new features like a password strength indicator or additional validation becomes straightforward.
> * **Reusability**: Common functionality like email validation or hashing can be reused in other forms (e.g., login or password reset).
> * **Testability**: Each module (UI, logic, utilities, and API) can be tested independently.
> * **Readability**: By separating concerns, it becomes easier for developers to understand and maintain the code.
> 
> ### **4. Example Implementation**
> You can refer to the following structure as an example:
> 
> * `SignupForm.tsx`: Focuses on rendering the form and integrating sub-components.
> * `useSignupForm.ts`: Centralizes state management and business logic.
> * `api.ts`: Handles API-related code.
> * `cryptoUtils.ts`: Contains utility functions for cryptographic and validation operations.

**Comment by keya5406** [comment]
> Well done implementing the signup functionality! I like the form with shadow, the button, its colors & font. 🆒 👍
> 
> Here are a few improvements: a) Add loader symbol after clicking any button till response is received from backend. b) I signed up but did not navigate to login page. Can you check the reason?
> 
> After these two improvements, I’d like to suggest a more industry-standard way of structuring the code to improve maintainability, readability, and scalability. Below are my observations and recommendations:
> 
> ### **1. Single Responsibility Principle**
> Currently, the `SignupForm` component handles multiple responsibilities:
> 
> * UI rendering
> * State management
> * Validation logic
> * API calls
> * Cryptographic operations
> 
> This makes the code harder to test, debug, and scale. To adhere to the single-responsibility principle, we can break it down into smaller, focused modules.
> 
> ### **2. Proposed Code Structure**
> Here’s how the code can be modularized:
> 
> 1. **Component (`SignupForm.tsx`)**
>    The `SignupForm` component should focus only on rendering the form and sub-components (e.g., `Email`, `Password`, `Button`) and delegate the state and logic to a custom hook (`useSignupForm`).
> 2. **Custom Hook (`useSignupForm.ts`)**
>    The hook manages form state, handles input changes, and executes the signup flow. It keeps the form logic isolated from the UI, making it easier to test independently.
> 3. **API Integration (`api.ts`)**
>    Extract API calls into a separate file to isolate external dependencies. This helps in mocking and testing API interactions separately.
> 4. **Utility Functions (`cryptoUtils.ts`)**
>    Move cryptographic operations (`generateSalt`, `hashPassword`) and validation logic (`isValidEmail`, `passwordPattern`) to a utility file. These can be reused across other parts of the application.
> 
> ### **3. Advantages of This Approach**
> * **Scalability**: Adding new features like a password strength indicator or additional validation becomes straightforward.
> * **Reusability**: Common functionality like email validation or hashing can be reused in other forms (e.g., login or password reset).
> * **Testability**: Each module (UI, logic, utilities, and API) can be tested independently.
> * **Readability**: By separating concerns, it becomes easier for developers to understand and maintain the code.
> 
> ### **4. Example Implementation**
> You can refer to the following structure as an example:
> 
> * `SignupForm.tsx`: Focuses on rendering the form and integrating sub-components.
> * `useSignupForm.ts`: Centralizes state management and business logic.
> * `api.ts`: Handles API-related code.
> * `cryptoUtils.ts`: Contains utility functions for cryptographic and validation operations.

 Yes, I will add a loader symbol after clicking any button, which will remain visible until the response is received from the backend.

 For proper navigation to the login page, the Login component is currently in a separate branch. You can check the PMUI-9-User-Login branch to review the navigation logic to the login page.

**Review by chintal31 — APPROVED** [review]
LGTM! 🎉

**Pull request merged**